### ISSUE TITLE
Contiguous KV-cache implementation of AttentionBackend using torch-spyre

### ISSUE THREAD

**Issue by tdoublep** [torch-spyre/spyre-inference#122]
Some simple attention implementations are already supported using torch-spyre but have certain limitations (e.g., contiguous KV cache is assumed).

This issue involves building a first, 0-th order, attention backend in the new stack. We assume that the backend will "gather" the discontiguous pages from GPU memory into a contiguous memory region before calling any attention kernel. Any additional constraints will be worked around in a similar manner.

This is just a stepping stone until we have a paged attention implementation working, but is a valuable as an initial "proof of life".

**Comment by kiszk** [comment]
In my understanding, when we use [Static cache](https://huggingface.co/docs/transformers/kv_cache#fixed-size-cache), it requires ["scatter"](https://github.com/huggingface/transformers/blob/main/src/transformers/cache_utils.py#L339-L340). This is the very first `-1`-th order, attention for HuggingFace, not for vLLM. Now, we are working for implementing `index_copy_` (i.e. scatter).

Then, we will work for "gather" for the `0`-th order.

What do you think?

**Comment by bringlein** [comment]
Based on discussions in torch-spyre/spyre-inference#123, the next step is to create a contiguous pytorch-native attention backend in vLLM that can be compiled by torch-spyre already now. 
CC @jvlunteren @bohnstingl

**Comment by kiszk** [comment]
Sounds good. In my understanding, "a contiguous pytorch-native attention backend in vLLM" is very similar to [Static cache](https://huggingface.co/docs/transformers/kv_cache#fixed-size-cache) in HF. Correct?

**Comment by bringlein** [comment]
> "a contiguous pytorch-native attention backend in vLLM" is very similar to Static cache in HF. Correct?

No, I'm not sure this is correct. We mean by "contiguous pytorch-native attention backend" a function that copies the paged cache into one continuous cache for every attention call. So it respects the paging of vLLM. 

Yes, this will have horrible performance (and twice the memory usage), but in this issue we want to get the functionality working first. Performance optimizations then come later (e.g. torch-spyre/spyre-inference#121 ), based on the experiences we gain here.

**Comment by kiszk** [comment]
Thank you for the clarification.

Can I confirm this, too? Will this operation, "copies the paged cache into one continuous cache for every attention call", be executed on host (i.e. CPU)?

**Comment by bohnstingl** [comment]
I have worked on a naive PyTorch implementation for the attention backend and put the first draft version [here](https://github.com/bohnstingl/vllm/tree/naive_attn_backend).

The way it currently works is:
1. It uses the KV cache of vLLM in its forward function, [here](https://github.com/bohnstingl/vllm/blob/61060fd27174a9b682b507351065f4d5ba5b68ce/vllm/v1/attention/backends/pytorch_native_attn.py#L232-L242)
2. It updates the KV cache by writing to the specific locations determined by the `block_table`, [here](https://github.com/bohnstingl/vllm/blob/61060fd27174a9b682b507351065f4d5ba5b68ce/vllm/v1/attention/backends/pytorch_native_attn.py#L255-L261). This operation is not explicitly a `scatter_` operation, but it writes to a specific location of a tensor, thus it might lower to a `scatter` operation.
3. It generates a block mask [here](https://github.com/bohnstingl/vllm/blob/61060fd27174a9b682b507351065f4d5ba5b68ce/vllm/v1/attention/backends/pytorch_native_attn.py#L268-L277). This is currently a bit cumbersome and awkwardly done with a set of loop, but it can definitely be improved to be more efficient. I think in the first iteration this mask generation might even run on `cpu`
4. It performs the actual attention computation [here](https://github.com/bohnstingl/vllm/blob/61060fd27174a9b682b507351065f4d5ba5b68ce/vllm/v1/attention/backends/pytorch_native_attn.py#L288-L294), only using torch native ops, such as `repeat_interleave`, `transpose`, `matmul`, `mul` and `softmax`.
5. It fetches the relevant outputs [here](https://github.com/bohnstingl/vllm/blob/61060fd27174a9b682b507351065f4d5ba5b68ce/vllm/v1/attention/backends/pytorch_native_attn.py#L297-L300), which is basically a slicing operation, which again might be lowered to a `gather`.
6. It returns the final outputs

Ideally, the KV cache tensor remains on the spyre device and gets updated in-place.

@kiszk what do you think about the necessary operations in this implementation? Would they be supported already? I think one blocker currently is the support of matmul for 3D and 4D tensors, right?

cc @jvlunteren @tdoublep

**Comment by bringlein** [comment]
summary of our discussion: one parallel approach would be to build a softmax-based attention only with the already supported operations to run on spyre, and then expand from there.

**Comment by kiszk** [comment]
@bohnstingl Thank you for sharing an implementation. It helps us greatly. I found some challenges to torch-spyre. After further discussion on our end, we will get back to you.

**Comment by bohnstingl** [comment]
Just to clarify, the version that I put together should just serve as a starting point, a first working version in vLLM that we might want to build on top. There is no guarantees for performance, neither is every single operation strictly necessary to be done in that way. For example, the `.item()` operations in the loop might be done in a different way if there are hardware limitations. Same for the mask generation and the attention computation itself.

**Comment by bohnstingl** [comment]
Another note, this code is actually closer to a "paged attention" related to https://github.com/vllm-project/vllm-spyre/issues/648.

One could try to use it as a basis to implement the simpler contiguous version for this issue here.

**Comment by jvlunteren** [comment]
I created a PR (https://github.com/vllm-project/vllm-spyre/pull/771) that adds a stand‑alone test script to facilitate developing a native PyTorch‑based attention kernel that runs on Spyre. It replaces the existing attention computation in https://github.com/bohnstingl/vllm/tree/naive_attn_backend with a simplified Spyre‑compatible variant derived from [deleted] The script compares the original implementation (`_compute_attention1`) with the Spyre‑compiled version (`_compute_attention2`), and after several adjustments, both now produce matching outputs.

**Comment by tdoublep** [comment]
Linking to the corresponding torch-spyre issues: 
- https://github.com/torch-spyre/torch-spyre/issues/424
- https://github.com/torch-spyre/torch-spyre/issues/911

**Comment by bringlein** [comment]
update: 
- we want to merge the PR
- and minimize part on CPU (chunk processing etc) as next step
- it is ok to restrict the possible usage scenarios for this for the time being

**Comment by tdoublep** [comment]
Tom will try to review this week + merge
